src/librustc/mir/interpret/error.rs

```diff
@@ -304,3 +303,41 @@ impl<'tcx> fmt::Display for EvalError<'tcx> {
         }
     }
 }
+
+#[derive(Debug, PartialEq, Eq, Clone, RustcEncodable, RustcDecodable)]
+pub enum ConstMathErr {
+    Overflow(Op),
+    DivisionByZero,
+    RemainderByZero,
+}
+pub use self::ConstMathErr::*;
+
+#[derive(Debug, PartialEq, Eq, Clone, RustcEncodable, RustcDecodable)]
+pub enum Op {
```

Comment:
You can just reuse `mir::BinOp` and have another variant for negation (just like division by zero has its own variant). Also, you can move these variants to `EvalErrorKind`.

Reply:
BinOp has more variants, so I'd have to reinsert the `unreachable!` branch or fill out error descriptions for things that'll never happen

Reply:
Also the `ConstMathErr` type is used elsewhere (e.g. mir assertions). Do we really want to allow all possible `EvalErrorKind`s in mir assertions?

Reply:
I guess you could rename `ConstMathErr` to `ArithErr` and `Op` to `ArithOp`.

Reply:
That or just inline `Op` into the error enum, since I don't think it's used elsewhere (so e.g. `AddOverflow`).

Reply:
I didn't miss these comments, I just wanted to see how far it goes, and simply using `EvalErrorKind` everywhere makes a lot of code much simpler. Should I undo it and keep `ConstMathErr`/`ArithErr`?

Reply:
Yeah, just because we can get a simple `&str` from them, and it's a finite set of things that we can generate builtin panics for. Unleessss we can turn `Assert` into something we can stash `miri` errors into! E.g.:
https://github.com/rust-lang/rust/blob/88cd36714c6d62393f18d8efa2ea2724e1ec4a9d/src/librustc/mir/interpret/error.rs#L63
could be used instead of:
https://github.com/rust-lang/rust/blob/88cd36714c6d62393f18d8efa2ea2724e1ec4a9d/src/librustc/mir/mod.rs#L1192-L1195
by "simply" having `EvalErrorKind` parametrized on some `O` which would be `u64` (by default / everywhere) except in `mir::TerminatorKind::Assert` where it would be `Operand<'tcx>`.
One downside of it is you have to write MIR visiting code for `EvalErrorKind` but it shouldn't be *too* bad.

Reply:
If you don't like that *crazy* idea, I prefer `ArithErr`.